### PR TITLE
test: migrate client module api tests to junit 5

### LINKED ISSUE #4322: Migrate tests from junit4 to junit5
JUnit 5 offers significant improvements over JUnit 4, including more powerful and flexible testing capabilities like annotation-based coding(so we don't need to write assume), parallel test execution, and other features. We can write new test classes using junit5, and existing-test classes can be slowly migrated.

This issue to track tests migrate to junit5, we can migrate a package、a small module gradually.

See also:
 https://lists.apache.org/thread/hlyc722gy45s3y61q7n79hdcr6cgrx7v

### PR DESCRIPTION
**Pull request by sherlock-lin**
Descriptions of the changes in this PR:

Fix #4322

Main Issue: #4322

### CODE REVIEW

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/extension/TestContextExtension.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ */
+package org.apache.bookkeeper.client.extension;
+
+import org.junit.jupiter.api.extension.BeforeEachCallback;
+import org.junit.jupiter.api.extension.ExtensionContext;
+
+/**
+ * Unit Test Context Information.
+ */
+public class TestContextExtension implements BeforeEachCallback {
```
Comment (shoothzj) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/extension/TestContextExtension.java:27]: I think we can put this common class into `testtools`, WDYT?

Comment (sherlock-lin) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/extension/TestContextExtension.java:27]: > I think we can put this common class into `testtools`, WDYT?

Of course, I would also like this class to become more generalized, how do I add it to testtools please?

Comment (lhotari) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/extension/TestContextExtension.java:27]: This class isn't needed. Junit5 already contains [TestInfo](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/TestInfo.html) for this purpose. 
Please check https://junit.org/junit5/docs/current/user-guide/#writing-tests-dependency-injection for the example.

**Review by lhotari — CHANGES_REQUESTED** [review]
Use JUnit5's TestInfo instead of adding custom extension.

**Comment by shoothzj** [comment]
@lhotari Thanks for you remind, I remember this class too, and it has been used in https://github.com/apache/bookkeeper/blob/5f73147a2803a5147d9d9ba2d28eaa6c79c998a3/bookkeeper-server/src/test/java/org/apache/bookkeeper/test/BookKeeperClusterTestCase.java#L178-L180

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/api/BookKeeperBuildersOpenLedgerTest.java [hunk 2]
```diff
@@ -58,20 +56,20 @@ public class BookKeeperBuildersOpenLedgerTest extends MockBookKeeperTestCase {
...
     public static Collection<Object[]> data() {
         return Arrays.asList(new Object[][]{
                 {true},
                 {false}
         });
     }
 
-    @Test
-    public void testOpenLedger() throws Exception {
+    @MethodSource("data")
```
Comment (lhotari) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/api/BookKeeperBuildersOpenLedgerTest.java:70]: In junit5, it's better to use `@ValueSource(booleans =  {true, false})` for cases where there's a single boolean parameter.

Comment (sherlock-lin) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/api/BookKeeperBuildersOpenLedgerTest.java:70]: @lhotari Thanks，It's a better choice.

**Review by hezhangjian — APPROVED** [review]
minor nits: I think **volatile** in `private volatile String testName;` might be useless

**Comment by sherlock-lin** [comment]
> minor nits: I think **volatile** in `private volatile String testName;` might be useless
done  :)

**Comment by sherlock-lin** [comment]
> minor nits: I think **volatile** in `private volatile String testName;` might be useless

done :)

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/MockBookKeeperTestCase.java [hunk 3]
```diff
@@ -264,6 +267,7 @@ private DigestManager getDigestType(long ledgerId) throws GeneralSecurityExcepti
                 UnpooledByteBufAllocator.DEFAULT, false);
     }
 
+    @AfterEach
     @After
```
Comment (liangyuanpeng) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/MockBookKeeperTestCase.java:271]: how about add some todo here: `// TODO: delete @Before when only run with junit5`

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/MockBookKeeperTestCase.java [hunk 4]
```diff
@@ -141,6 +143,7 @@ public MockEntry(byte[] payload, long lastAddConfirmed) {
 
     }
 
+    @BeforeEach
     @Before
```
Comment (liangyuanpeng) [inline bookkeeper-server/src/test/java/org/apache/bookkeeper/client/MockBookKeeperTestCase.java:147]: how about add some todo here: `// TODO: delete @Before when only run with junit5`

**Comment by sherlock-lin** [comment]
Help with merge，please

**Comment by sherlock-lin** [comment]
@lhotari  help review，please

**Review by lhotari — APPROVED** [review]
LGTM

**Comment by lhotari** [comment]
Good work @sherlock-lin

**Comment by lhotari** [comment]
I'll close and reopen to see if it passes with latest changes in master.

File: bookkeeper-server/src/test/java/org/apache/bookkeeper/client/MockBookKeeperTestCase.java (hunk 3 above)
Comment (sherlock-lin) [reply bookkeeper-server/src/test/java/org/apache/bookkeeper/client/MockBookKeeperTestCase.java:271]: https://github.com/apache/bookkeeper/issues/4491

**Comment by shoothzj** [comment]
Thanks for your contribution :)

**Pull request merged**